google-cloud-spanner/lib/google/cloud/spanner/pool.rb

```diff
@@ -27,49 +29,113 @@ module Spanner
         end
 
         def with_session
-          session = checkout
-          yield session
-          checkin session
+          session = checkout_session
+          begin
+            yield session
+          ensure
+            checkin_session session
+          end
+        end
+
+        def checkout_session
+          @mutex.synchronize do
+            loop do
+              read_session = session_queue.shift
+              return read_session if read_session
+              write_transaction = transaction_queue.shift
+              return write_transaction.session if write_transaction
+
+              return new_session! if can_allocate_more_sessions?
+
+              fail SessionLimitError if @fail
+
+              @resource.wait @mutex
+            end
+          end
+        end
+
+        def checkin_session session
+          unless all_sessions.include? session
+            fail ArgumentError, "Cannot checkin session"
+          end
+
+          @mutex.synchronize do
+            session_queue.push session
+
+            @resource.broadcast
+          end
+
+          nil
         end
 
-        def checkout
-          if queue.empty?
-            fail "No available sessions" if pool.size >= @max
-            new_session!
+        def with_transaction
+          tx = checkout_transaction
+          begin
+            yield tx
+          ensure
+            checkin_transaction tx
           end
-          queue.shift
         end
 
-        def checkin session
-          fail "Cannot checkin session" unless pool.include? session
+        def checkout_transaction
+          @mutex.synchronize do
+            loop do
+              write_transaction = transaction_queue.shift
+              return write_transaction if write_transaction
+              read_session = session_queue.shift
+              return read_session.create_transaction if read_session
+
+              return new_transaction! if can_allocate_more_sessions?
 
-          # Do we ever delete sessions if the queue is *too* full?
-          queue.push session
+              fail SessionLimitError if @fail
+
+              @resource.wait @mutex
+            end
+          end
+        end
+
+        def checkin_transaction tx
+          unless all_sessions.include? tx.session
+            fail ArgumentError, "Cannot checkin session"
+          end
+
+          @mutex.synchronize do
+            transaction_queue.push tx.session.create_transaction
+
+            @resource.broadcast
+          end
 
           nil
         end
 
         def keepalive!
           ensure_keepalive_thread!
 
-          # Call keep alive only on the queue, not all sessions.
-          # If a session is checked out, we can assume its being used.
-          queue.each(&:keepalive!)
+          # Call keep alive only on the queue, not all sessions or transactions.
+          # If a session or transaction is checked out assume its being used.
+          session_queue.each(&:keepalive!)
```

Comment:
I dont think it makes sense to do keepalive on all the sessions in the queue. Instead you should keep track of when the session was last used and only if the last use time was more than 30 minutes in the past should keep alive be called.